### PR TITLE
Validate the recurrence schedule fields and set defaults

### PR DESCRIPTION
**Pull request by bmuniswamy**
# Critical Changes

# Changes

# Issues Closed

# New Metadata

# Deleted Metadata

# Definition of Done
  Refer to [Asteroids DoD document](https://salesforce.quip.com/iq2mAy4i62oM) to see any additional details for the items below
- [x] Place holder data is incorporated (Refer to [Case Management Sample Data document](https://quip.com/cbFcAPJF8t0z))
- [ ] Base axe-core JEST test included for any new LWC (No critical violations)
- [ ] CRUD/FLS is enforced in Apex
- [ ] Update & Test permission sets to account for new features and CRUD/FLS
- [ ] All modified classes are unit tested (Apex) at 100% coverage
- [ ] UX approval or UX not necessary
- [ ] PR contains draft release notes
- [x] Attach work item to the pull request with [Lurch](https://salesforce.quip.com/50ZRA5LEWVzH) `**Lurch:attach W-000000` or `**Lurch:add`
- [ ] All **acceptance criteria** have been met
    - [x] Developer
    - [ ] Code Reviewer
    - [ ] QA
- [ ] QE story level testing completed

### CODE REVIEW

**Comment by bmuniswamy** [comment]
[W-8124256](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B00000089hByIAI/view)

**Review by lbeller — CHANGES_REQUESTED** [review]
Request changing label "At least one day of the week must be selected when Weekly is selected." to "Select at least one day of the week."

File: force-app/main/default/classes/ServiceScheduleService.cls [hunk 1]
```diff
@@ -94,9 +94,14 @@ public with sharing class ServiceScheduleService {
     }
 
     private List<DateTime> getDateTimes(ServiceSchedule__c schedule) {
+        if (schedule.Frequency__c == 'OneTime') {
```
Comment (jjbennett) [inline force-app/main/default/classes/ServiceScheduleService.cls:97]: Should 'One Time' be a picklist value? (Note the space)

File: force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js [hunk 2]
```diff
@@ -49,13 +56,74 @@ export default class NewServiceSchedule extends LightningElement {
...
+            errMessages.push(NO_END_LABEL);
+        }
+
+        if (!hasDayOfWeek) {
+            errMessages.push(DAY_REQUIRED_LABEL);
+        }
+
+        this.errorMessage = errMessages.join("\n");
+
+        this.isValid = isFormValid && datesValid && hasEndCondition && hasDayOfWeek;
```
Comment (jjbennett) [inline force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js:96]: I like how you broke all of these conditions down, it is easy to follow what is required with this one line here.

File: force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js [hunk 3]
```diff
@@ -49,13 +56,74 @@ export default class NewServiceSchedule extends LightningElement {
...
+
+        return this.isValid;
+    }
+
+    validateServiceScheduleOnOrAfter() {
+        let noError = true;
+
+        this.template.querySelectorAll("lightning-input-field").forEach(field => {
+            if (this.picklistFields.frequency.value !== ONE_TIME) {
```
Comment (jjbennett) [inline force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js:105]: We could return early at the top of this method when frequency is one time. If it is not one-time, could we just check this.picklistFields.seriesEndsOn value?

I'm not sure I'm following why we are querying all the fields and then the picklist fields here?

File: force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js [hunk 4]
```diff
@@ -153,6 +233,20 @@ export default class NewServiceSchedule extends LightningElement {
             (endTime.getMinutes() - startTime.getMinutes()) / 60;
     }
 
+    setFirstSessionStartTimeAndEndTime(startDate) {
+        if (!startDate) {
+            return;
```
Comment (jjbennett) [inline force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js:238]: Nice early return here :)

File: force-app/main/default/lwc/picklist/picklist.js [hunk 5]
```diff
@@ -22,9 +22,15 @@ export default class Picklist extends LightningElement {
         }
 
         this._options = this.picklist.picklistValues.map(picklistValue => {
-            // TODO: Set defaultValue if value not provided?
             let isSelected = this.value && this.value.includes(picklistValue.value);
 
+            if (
```
Comment (jjbennett) [inline force-app/main/default/lwc/picklist/picklist.js:27]: I'm not sure this is working as intended. If picklistValue.defaultValue and !this.value would we just want to set this.value to picklistValue.defaultValue?

**Review by jjbennett — COMMENTED** [review]
I had some questions around some of the conditionals but overall this is looking really great.

File: force-app/main/default/classes/ServiceScheduleService.cls (hunk 1 above)
Comment (bmuniswamy) [reply force-app/main/default/classes/ServiceScheduleService.cls:97]: Should the picklist value have a space. I was not aware so I did not add the space. I can add that if the team thinks that way

File: force-app/main/default/lwc/picklist/picklist.js (hunk 5 above)
Comment (bmuniswamy) [reply force-app/main/default/lwc/picklist/picklist.js:27]: I think I was testing something and left it in. Sorry I removed it. Thanks for catching it

File: force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js (hunk 3 above)
Comment (bmuniswamy) [reply force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js:105]: Omg.. I am not sure why I overcomplicated things here.. That one thing would do it..

File: force-app/main/default/lwc/picklist/picklist.js [hunk 6]
```diff
@@ -22,9 +22,12 @@ export default class Picklist extends LightningElement {
         }
 
         this._options = this.picklist.picklistValues.map(picklistValue => {
-            // TODO: Set defaultValue if value not provided?
             let isSelected = this.value && this.value.includes(picklistValue.value);
 
+            if (isSelected && this.value === picklistValue.value) {
+                this.value = picklistValue.value;
```
Comment (jjbennett) [inline force-app/main/default/lwc/picklist/picklist.js:28]: I think this might still not be what you were going after, let's discuss Monday.

File: force-app/main/default/classes/ServiceScheduleService.cls (hunk 1 above)
Comment (jjbennett) [reply force-app/main/default/classes/ServiceScheduleService.cls:97]: We need to update these value checks, maybe a constant on the service would work?

File: force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js [hunk 7]
```diff
@@ -49,13 +56,46 @@ export default class NewServiceSchedule extends LightningElement {
...
+            errMessages.push(
+                format(START_BEFORE_END_LABEL, [
+                    this.dateFields.start.label,
+                    this.dateFields.end.label,
+                ])
+            );
+        }
+
+        if (!hasEndCondition) {
+            errMessages.push(NO_END_LABEL);
```
Comment (jjbennett) [inline force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js:87]: This appears to be misfiring.

File: force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js [hunk 8]
```diff
@@ -98,6 +137,18 @@ export default class NewServiceSchedule extends LightningElement {
...
     }
 
+    handleLoad() {
+        this.template.querySelectorAll("lightning-input-field").forEach(field => {
+            if (field.fieldName === this.dateFields.start.apiName) {
+                this.setFirstSessionStartTimeAndEndTime(field.value);
+            }
+        });
+
+        this.picklistFields.frequency.value = !this.picklistFields.frequency.value
```
Comment (jjbennett) [inline force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js:147]: Is this still necessary now that you are respecting defaultValues in the picklist component?

Comment (bmuniswamy) [reply force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js:147]: Yes unfortunately it is necessary else I get an error message saying unable to de-reference a null object in the console since the value is being passed as null

File: force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js [hunk 9]
```diff
@@ -153,7 +206,39 @@ export default class NewServiceSchedule extends LightningElement {
...
+
+    validateServiceScheduleOnOrAfter() {
+        let noError = true;
+
+        if (this.picklistFields.frequency.value === ONE_TIME) {
+            return noError;
+        }
+
+        noError =
+            (this.template.querySelector('[data-element="seriesEndsOn"]') === null ||
```
Comment (jjbennett) [inline force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js:235]: This looks great but these large conditionals can sometimes be hard to follow. What do you think about breaking this down, maybe checking the value of  the seriesEnd picklist field and if it is ON then make sure seriesEndsOn has a value else make sure that number of sessions has a value?

Comment (bmuniswamy) [reply force-app/main/default/lwc/newServiceSchedule/newServiceSchedule.js:235]: I had done it that way but I thought you would not like it so did it this way. I can change it

**Review by jjbennett — APPROVED** [review]
Thanks for iterating on this. Great work!

**Review by lbeller — APPROVED** [review]
